core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1285,13 +1290,35 @@ case class GroupConfig(groupMinSessionTimeoutMs: Int,
 case class JoinGroupResult(members: List[JoinGroupResponseMember],
                            memberId: String,
                            generationId: Int,
-                           subProtocol: String,
+                           protocolType: Option[String],
+                           protocolName: String,
                            leaderId: String,
                            error: Errors)
 
-case class SyncGroupResult(memberAssignment: Array[Byte],
+object JoinGroupResult {
+  def apply(memberId: String, error: Errors): JoinGroupResult = {
+    JoinGroupResult(
+      members = List.empty,
+      memberId = memberId,
+      generationId = GroupCoordinator.NoGeneration,
+      protocolType = None,
+      protocolName = GroupCoordinator.NoProtocol,
```

Comment:
The inconsistency with SyncGroup is a little annoying, but I guess we have to keep it for compatibility. I was considering if it's worthwhile making the this field nullable going forward.

Reply:
You bring a really good point that I haven't considered. I have been looking at it and I think that it makes sense to make it nullable going forward. It means that Protocol Name would be null in case of an error instead of an empty string for versions 7 and above. Overall, it makes the internals better, the API more consistent and the changes in the API layer to handle the backward compatibility are small so I am up for it.

I went ahead and made the changes, let me know what do you think.

Reply:
From a non OO language perspective, changing the protocol name field in the join group response from non-null to nullable isn't the greatest / is backwards incompatible.

Effectively, in Go, either I consider the `string` type in protocol to be a legitimate object that can at some future point be nullable, or I use the Go `string` primitive and hope that the Java side never changes the equation.

This will be the first instance of a primitive (non array) switching from non-nullable to nullable. So, while from a backwards compatibility perspective, there are no problems for the Java client, there are problems for clients of non-OO languages.

Because of the incompatibility, I can add support for decoding null as non-null, but I can't really add support for allowing the field itself in the JoinGroupResponse struct to be null. To do so would effectively necessitate adding a second field (non-null field v6 and below, nullable field v7 up).

Reply:
Thank you for your feedback. You bring a point that we haven't considered. I think that it is more challenging for languages where the `string` type is not an object such a Go.

Have you considered changing the field type to a pointer to a string? Usually, the internal protocol is not exposed to the users of the consumer thus changing the type should only impact the internals which seems acceptable. That being said, I am not aware of your project so I may be completely wrong.

Reply:
I think switching to a pointer now is alright. I agree about the internal protocol not usually being exposed, so it's probably alright to have more backwards incompatible changes in that. The API of the protocol is public, though, to theoretically allow anybody to build on top of it.

[It's a work in progress](https://github.com/twmb/kafka-go/blob/master/pkg/kmsg/generated.go), but barring integration tests showing clear failures, the API is complete.